pkg/servicemesh/control-plane.go

```diff
@@ -0,0 +1,493 @@
+	apiClient *clients.Settings
+}
+
+// NewControlPlaneBuilder method creates new instance of builder.
+func NewControlPlaneBuilder(apiClient *clients.Settings, name, nsname string) *ControlPlaneBuilder {
+	glog.V(100).Infof("Initializing new serviceMeshControlPlane ControlPlaneBuilder structure with the following "+
+		"params: name: %s, namespace: %s", name, nsname)
+
+	builder := &ControlPlaneBuilder{
+		apiClient: apiClient,
+		Definition: &istioV2.ServiceMeshControlPlane{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: nsname,
+			},
+		},
+	}
+
+	if name == "" {
+		glog.V(100).Infof("The name of the serviceMeshControlPlane is empty")
+
+		builder.errorMsg = "The serviceMeshControlPlane 'name' cannot be empty"
+	}
+
+	if nsname == "" {
+		glog.V(100).Infof("The namespace of the serviceMeshControlPlane is empty")
+
+		builder.errorMsg = "The serviceMeshControlPlane 'namespace' cannot be empty"
+	}
+
+	return builder
+}
+
+// WithAllAddonsDisabled disables all addons to the serviceMeshControlPlane.
+func (builder *ControlPlaneBuilder) WithAllAddonsDisabled() *ControlPlaneBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	enablement := false
+
+	glog.V(100).Infof(
+		"Creating serviceMeshControlPlane %s with the all addons disabled", builder.Definition.Name)
+
+	disableAllAddons := istioV2.AddonsConfig{
+		Prometheus: &istioV2.PrometheusAddonConfig{
+			Enablement: istioV2.Enablement{Enabled: &enablement},
+		},
+		Grafana: &istioV2.GrafanaAddonConfig{
+			Enablement: istioV2.Enablement{Enabled: &enablement},
+		},
+		Kiali: &istioV2.KialiAddonConfig{
+			Enablement: istioV2.Enablement{Enabled: &enablement},
+		},
+		ThreeScale: &istioV2.ThreeScaleAddonConfig{
+			Enablement: istioV2.Enablement{Enabled: &enablement},
+		},
+	}
+
+	builder.Definition.Spec.Addons = &disableAllAddons
+
+	return builder
+}
+
+// WithGrafanaAddon adds grafana addon to the serviceMeshControlPlane.
+func (builder *ControlPlaneBuilder) WithGrafanaAddon(
+	enablement bool,
+	grafanaInstallConfig *istioV2.GrafanaInstallConfig,
+	address string) *ControlPlaneBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Creating serviceMeshControlPlane %s with the Grafana addons defined: enablement %s, "+
+			"grafanaInstallConfig %s, address %s", builder.Definition.Name, enablement, grafanaInstallConfig, address)
+
+	if enablement {
+		if grafanaInstallConfig == nil {
+			glog.V(100).Infof("The grafanaInstallConfig of the Grafana addon is empty")
+
+			builder.errorMsg = "The Grafana addon 'grafanaInstallConfig' cannot " +
+				"be empty when Grafana addon is enabled"
+		}
+
+		if address == "" {
+			glog.V(100).Infof("The address of the Grafana addon is empty")
+
+			builder.errorMsg = "The Grafana addon 'address' cannot be empty when Grafana addon is enabled"
+		}
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	addons := &istioV2.AddonsConfig{
```

Comment:
Is there a use case for enabling multiple addons at once? If so, it seems like you currently overwrite any previous addons so only one can ever be added at a time. If you want to support multiple, I think the code needs to look like this:
```
addonConfig := &istioV2.GrafanaAddonConfig{
  Enablement: istioV2.Enablement{
    Enabled: &enablement,
  },
  Install: grafanaInstallConfig,
  Address: &address,
}

if builder.Definition.Spec.Addons == nil {
    builder.Definition.Spec.Addons = new(istioV2.AddonsConfig)
}

builder.Definition.Spec.Addons.Grafana = addonConfig

return builder
```